src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Tree/Dependencies/DependencyTreeTelemetryService.cs

```diff
@@ -0,0 +1,94 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.ComponentModel.Composition;
+using System.Linq;
+using Microsoft.VisualStudio.ProjectSystem.LanguageServices;
+using Microsoft.VisualStudio.Telemetry;
+
+namespace Microsoft.VisualStudio.ProjectSystem.VS.Tree.Dependencies
+{
+    /// <summary>
+    /// Maintain light state to fire telemetry on dependency tree
+    /// </summary>
+    [Export(typeof(IDependencyTreeTelemetryService))]
+    [AppliesTo(ProjectCapability.DependenciesTree)]
+    internal class DependencyTreeTelemetryService : IDependencyTreeTelemetryService
+    {
+        private const string TelemetryEventName = "TreeUpdated";
+        private const string UnresolvedLabel = "Unresolved";
+        private const string ResolvedLabel = "Resolved";
+
+        private readonly IProjectAsynchronousTasksService _tasksService;
+        private readonly ITelemetryService _telemetryService;
+        private readonly Dictionary<string, bool> observedRuleChanges = new Dictionary<string, bool>();
+        private bool observedDesignTime = false;
+        private bool stopTelemetry = false;
+
+        [ImportingConstructor]
+        public DependencyTreeTelemetryService(
+            IUnconfiguredProjectCommonServices commonServices,
+            [Import(ExportContractNames.Scopes.UnconfiguredProject)]IProjectAsynchronousTasksService tasksService,
+            ITelemetryService telemetryService)
+        {
+            _tasksService = tasksService;
+            _telemetryService = telemetryService;
+        }
+
+        public void ObserveTreeUpdateCompleted()
+        {
+            if (stopTelemetry) return;
+
+            var notReady = !observedRuleChanges.Any() || !observedRuleChanges.All(entry => entry.Value);
+
+            _telemetryService.PostEvent($"{TelemetryEventName}/{(notReady ? UnresolvedLabel : ResolvedLabel)}");
```

Comment:
> _telemetryService.PostEvent($"{TelemetryEventName}/{(notReady ? UnresolvedLabel : ResolvedLabel)}"); [](start = 11, length = 101)

One suggestion here: we can also add elapsed time to this event don't we? At least in between resolved and unresolved etc. This would give use several telemetry events with elapsed time in between them and give idea about how long we usually load the tree. Like 1st unresolved elapsed=diff from the time when DependencyTreeTelemetryService was constructed; 2nd unresolved with elapsed with diff from 1st Unresolved and the finally Resolved event with diff from 2nd unresolved.  Since they all come from the same session we could potentially query and sum them later when analyzing telemetry.

Reply:
I think elapsed time is automatically added to the telemetry event. I see it in the Telemetry Monitor

---
In reply to: [145188931](https://github.com/dotnet/project-system/pull/2879#discussion_r145188931) [](ancestors = 145188931)